File: src/haven/purus/pbot/api/PBotGob.java
package haven.purus.pbot.api;

import haven.*;
import haven.Composite;
import haven.purus.BoundingBox;
import haven.purus.GobColor;
import haven.purus.GobText;
import haven.purus.pathfinder.Pathfinder;
import haven.render.MixColor;
import haven.render.sl.Array;

import java.awt.Color;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.Semaphore;

public class PBotGob {

	private final Gob gob;
	private PBotSession pBotSession;

	public static HashMap<String, String> gobWindowMap = new HashMap<String, String>() {{
		put("gfx/terobjs/crate", "Crate");
		put("gfx/terobjs/dframe", "Frame");
		put("gfx/terobjs/kiln", "Kiln");
		put("gfx/terobjs/fineryforge", "Finery Forge");
		put("gfx/terobjs/steelcrucible", "Steelbox");
		put("gfx/terobjs/smelter", "Ore Smelter");
		put("gfx/terobjs/pow", "Fireplace");
		put("gfx/terobjs/oven", "Oven");
		put("gfx/terobjs/cauldron", "Cauldron");
		put("gfx/terobjs/woodbox", "Woodbox");
		put("gfx/terobjs/create", "Crate");
		put("gfx/terobjs/furn/table-stone", "Table");
		put("gfx/terobjs/furn/cottagetable", "Table");
		put("gfx/terobjs/wbasket", "Basket");
		put("gfx/terobjs/chickencoop", "Chicken Coop");
		put("gfx/terobjs/htable", "Herbalist Table");
		put("gfx/terobjs/studydesk", "Study Desk");
		put("gfx/terobjs/cupboard", "Cupboard");
		put("gfx/terobjs/ttub", "Tub");
		put("gfx/terobjs/chest", "Chest");
	}};


	public PBotGob(Gob gob, PBotSession pBotSession) {
		this.pBotSession = pBotSession;
		this.gob = gob;
	}

	/**
	 * Click the gob
	 * @param btn 1 = left, 2 = middle, 3 = right
	 * @param mod Key modifier mask 1 = shift 2 = ctrl 4 = alt
	 * @param meshid can be a door, roasting spit etc.
	 * @param olid gob overlay to click, for example roasting spit
	 */
	public void doClick(int btn, int mod, int meshid, int olid) {
		pBotSession.gui.map.wdgmsg("click", pBotSession.PBotUtils().getCenterScreenCoord(), gob.rc.floor(OCache.posres), btn, mod, 1, (int)gob.id, gob.rc.floor(OCache.posres), olid, meshid);
	}

	/**
	 * Euclidean distance between this and target gob
	 * @param gob target
	 * @return the distance
	 */
	public double dist(PBotGob gob) {
		return gob.getCoords().dist(getCoords());
	}

	/**
	 * Returns rc-coords of the gob
	 * @return Coord object with x and y attributes
	 */
	public Coord2d getCoords() {
		final Coord2d[] ret = {null};
		Semaphore s = new Semaphore(1);
		s.acquireUninterruptibly();
		new Runnable() {
			@Override
			public void run() {
				try {
					ret[0] = Coord2d.z;
					ret[0] = new Coord2d(gob.getc());
				} catch(MCache.LoadingMap lm) {
					if(ret[0] == null) {
						lm.waitfor(this, waiting -> {
						});
						return;
					}
				}
				s.release();
			}
		}.run();
		s.acquireUninterruptibly();
		return ret[0];
	}

	/**
	 * Get gob id of the gob (not unique through sessions)
	 * @return id
	 */
	public long getId() {
		return this.gob.id;
	}

	/**
	 * Check if stockpile is full
	 * @return True if stockpile is full, else false
	 */
	public boolean stockpileIsFull() {
		return gob.getattr(ResDrawable.class).sdt.peekrbuf(0) == 31;
	}

	/**
	 * Itemact with gob, to fill trough with item in hand for example
	 * @param mod 1 = shift, 2 = ctrl, 4 = alt
	 */
	public void itemClick(int mod) {
		pBotSession.gui.map.wdgmsg("itemact", Coord.z, gob.rc.floor(OCache.posres), mod, 0, (int) gob.id, gob.rc.floor(OCache.posres), 0, -1);
	}

	/**
	 * Add cool hovering text above gob
	 * @param text text to add
	 * @param height height that the hext hovers at
	 * @return id id of text used to remove it
	 */
	public int addGobText(String text, int r, int g, int b, int a, int height) {
		Gob.Overlay gt = new Gob.Overlay(gob, new GobText(gob, text, new Color(r, g, b, a), height));
		pBotSession.gui.ui.sess.glob.loader.defer(() -> {synchronized(gob) {
			gob.addol(gt);
		}}, null);
		return gt.id;
	}

	/**
	 * Remove an added hovering text from gob that was added with addGobText
	 * @param id Id of the gobtext
	 */
	public void removeGobText(int id) {
		pBotSession.gui.ui.sess.glob.loader.defer(() -> {synchronized(gob) {
			gob.findol(id).remove();
		}}, null);
	}
	/**
	 * Returns the name of the gobs resource file, or null if not found
	 * @return Name of the gob
	 */
	public String getResname() {
		while(true) {
			try {
				if(gob.getres() != null)
					return gob.getres().name;
				else
					return null;
			} catch(Loading l) {
				PBotUtils.sleep(10);
			}
		}
	}

	/**
	 * Get name of window for gob from gobWindowMap
	 * @return Inventory window name
	 */
	public String windowNameForGob() {
		while(true) {
			try {
				if(gob.getres() == null)
					return null;
				else
					return gobWindowMap.get(gob.getres().name);
			} catch(Loading l) {
				PBotUtils.sleep(10);
			}
		}
	}


	/**
	 * Set gob as colored, replaces previous if color already set
	 * @param r Red between 0-255
	 * @param g Green between 0-255
	 * @param b Blue between 0-255
	 * @param a Alpha between 0-255
	 */
	public void setMarked(int r, int g, int b, int a) {
		pBotSession.gui.ui.sess.glob.loader.defer(() -> {synchronized(gob) {
			gob.setattr(new GobColor(gob, new MixColor(r, g, b, a)));
		}}, null);
	}

	/**
	 * Remove color marking
	 */
	public void setUnmarked() {
		pBotSession.gui.ui.sess.glob.loader.defer(() -> {synchronized(gob) {
			gob.delattr(GobColor.class);
		}}, null);
	}

	/**
	 * Click a gob with pathfinder, with given button
	 * @param btn 1 = left click, 3 = right click
	 * @param mod 1 = shift, 2 = ctrl, 4 = alt
	 * @param meshId meshid to click
	 */
	public void pfClick(int btn, int mod, int meshId) {
		Pathfinder.run(gob.rc, gob, btn, mod, meshId,"", pBotSession.gui);
	}

	/**
	 * Click a gob with pathfinder, with given button
	 * @param btn 1 = left click, 3 = right click
	 * @param mod 1 = shift, 2 = ctrl, 4 = alt
	 * @param meshId meshid to click
	 * @param olid overlay id to click for example roasting spit
	 */
	public void pfClick(int btn, int mod, int meshId, int olid) {
		Pathfinder.run(gob.rc, gob, btn, mod, olid, meshId,"", pBotSession.gui);
	}

	/**
	 * Click a gob with pathfinder, with given button
	 * @param btn 1 = left click, 3 = right click
	 * @param mod 1 = shift, 2 = ctrl, 4 = alt
	 * @param action dig, destroy etc to select before the final click
	 */
	public void pfClick(int btn, int mod, List<String> action) {
		Pathfinder.run(gob.rc, gob, btn, mod, -1, String.join(" ", action), pBotSession.gui);
	}

	/**
	 * Click a gob with pathfinder, with given button
	 * @param btn 1 = left click, 3 = right click
	 * @param mod 1 = shift, 2 = ctrl, 4 = alt
	 */
	public void pfClick(int btn, int mod) {
		Pathfinder.run(gob.rc, gob, btn, mod, -1,"", pBotSession.gui);
	}

	/**
	 * Check if the object is moving
	 * @return Returns true if gob is moving, false otherwise
	 */
	public boolean isMoving() {
		if (gob.getv() == 0)
			return false;
		else
			return true;
	}

	/**
	 * Get speed of this gob if it is moving
	 * @return Speed of the gob, -1 if not moving object
	 */
	public double getSpeed() {
		LinMove lm = gob.getLinMove();
		if(lm == null)
			return -1;
		else
			return lm.getv();
	}

	/**
	 * Returns resnames of poses of this gob if it has any poses
	 * @return Resnames of poses
	 */
	public ArrayList<String> getPoses() {
		ArrayList<String> ret = new ArrayList<>();
		Drawable d = gob.getattr(Drawable.class);

		if(d instanceof Composite) {
			Composite comp = (Composite)d;
			for(ResData rd:comp.prevposes) {
				while(true) {
					try {
						ret.add(rd.res.get().name);
						break;
					} catch(Loading l) {
						PBotUtils.sleep(10);
					}
				}
			}
		}
		return ret;
	}

	/**
	 * Get overlays of the gob, get meshId with getOverlyId
	 * @return List containing resnames of the overlays
	 */
	public List<String> getOverlayNames() {
		List<String> ret = new ArrayList<>();
		for (Gob.Overlay ol : gob.ols) {
			if(ol.res != null)
				while(true) {
					try {
						ret.add(ol.res.get().name);
						break;
					} catch(Loading l) {
						PBotUtils.sleep(10);
					}
				}
		}
		return ret;
	}

	/**
	 * Return meshId of the overlay with given resname
	 * @param overlayName Exact match
	 * @return Meshid of the overlay -1 if not found
	 */
	public int getOverlayId(String overlayName) {
		for (Gob.Overlay ol : gob.ols) {
			if(ol.res != null)
			while(true) {
				try {
					if(ol.res.get().name.equals(overlayName)) {
						return ol.id;
					}
					break;
				} catch (Loading l) {
					PBotUtils.sleep(10);
				}
			}
		}
		return -1;
	}

	/**
	 * Sdt may tell information about things such as tanning tub state, crop stage etc.
	 * @return sdt of the gob, -1 if not found
	 */
	public int getSdt(int idx) {
		while(true) {
			try {
				Resource res = gob.getres();
				if(res != null) {
					GAttrib rd = gob.getattr(ResDrawable.class);
					return ((ResDrawable) rd).sdt.peekrbuf(idx);
				} else {
					return -1;
				}
			} catch(Loading l) {
				PBotUtils.sleep(10);
			}
		}
	}

	public long getGridId() {
		while(true) {
			try {
				return pBotSession.gui.ui.sess.glob.map.getgridt(getCoords().floor(MCache.tilesz)).id;
			} catch(MCache.LoadingMap l) {
				try {
					l.waitfor();
				} catch(InterruptedException e) {
					e.printStackTrace();
				}
			}
		}
	}

	// Get boundingbox of this gob, or null if not found
	public BoundingBox getBb() {
		return BoundingBox.getBoundingBox(gob);
	}

	public List<String> getCompEqu() {
		Composite comp = gob.getattr(Composite.class);
		List<String> ret = new ArrayList<>();
		if(comp != null) {
			if(comp.comp.cequ != null) {
				for(Composited.ED ed : comp.comp.cequ) {
					while(true) {
						try {
							ret.add(ed.res.res.get().name);
							break;
						} catch(Loading l) {
							try {
								l.waitfor();
							} catch(InterruptedException e) {
								e.printStackTrace();
							}
						}
					}
				}
			}
		}
		return ret;
	}

	public List<String> getCompMod() {
		Composite comp = gob.getattr(Composite.class);
		List<String> ret = new ArrayList<>();
		if(comp != null) {
			if(comp.comp.cmod != null) {
				for(Composited.MD md : comp.comp.cmod) {
					while(true) {
						try {
							ret.add(md.mod.get().name);
							break;
						} catch(Loading l) {
							try {
								l.waitfor();
							} catch(InterruptedException e) {
								e.printStackTrace();
							}
						}
					}
				}
			}
		}
		return ret;
	}

	/**
	 * Is the gob knocked out/dead
	 * @return True if animal is knocked out, false if not
	 */
	public boolean isKnocked() {
		return gob.knocked == Gob.Knocked.TRUE;
	}

}

File: scripts/__pbot/PBotGob.py
from typing import List, Tuple

class PBotGob(object):
    def __init__(self, gob):
        self._gob = gob

    ## Click the gob
    # @param btn 1 = left, 2 = middle, 3 = right
    # @param mod Key modifier mask 1 = shift 2 = ctrl 4 = alt
    # @param meshid can be a door, roasting spit etc.
    # @param olid gob overlay to click, for example roasting spit
    def do_click(self, btn: int = 1, mod: int = 0, meshid: int = -1, olid: int = 0):
        self._gob.doClick(btn, mod, meshid, olid)

    ## Click the gob with pathfinder
    # @param btn 1 = left, 2 = middle, 3 = right
    # @param mod Key modifier mask 1 = shift 2 = ctrl 4 = alt
    # @param meshid can be a door, roasting spit etc.
    # @param olid gob overlay to click, for example roasting spit
    def pf_click(self, btn: int = 1, mod: int = 0, meshid: int = -1, olid: int = 0):
        self._gob.pfClick(btn, mod, meshid, olid)

    ## Euclidean distance between this and target gob
    # @param gob target
    # @return the distance
    def dist(self, gob: "PBotGob") -> float:
        return float(self._gob.dist(gob._gob))

    ## Returns rc-coords of the gob
    # @return x and y coords
    def get_coords(self) -> Tuple[float, float]:
        c = self._gob.getCoords()
        return float(c.x), float(c.y),

    ## Get gob id of the gob
    # @return id
    def get_id(self) -> int:
        return int(self._gob.getId())

    ## Check if stockpile is full
    # @return true if stockpile is full else false
    def stockpile_is_full(self) -> bool:
        return self._gob.stockpileIsFull()

    ## Itemact with gob, to fill trough with item in hand for example
    # @param mod 1 = shift, 2 = ctrl, 4 = alt, combine bits for multiple
    def item_click(self, mod: int):
        self._gob.itemClick(mod)

    ## Add cool hovering text above the gob
    # @param text text to add
    # @param height height that the text hovers at
    # @param r red color 0-255
    # @param g green color 0-255
    # @param b blue color 0-255
    # @return id of text used to remove it
    def add_gob_text(self, text: str, height: int = 20, r: int = 255, g: int = 255, b: int = 255) -> int:
        return self._gob.addGobText(text, r, g, b, 255, height)

    ## Remove the added hovering text from gob that was added with add_gob_text
    # @param id id of the gobtext
    def remove_gob_text(self, id: int):
        self._gob.removeGobText(id)

    ## Returns the name of the gobs resource file or null if not found
    # @return gob resname
    def get_resname(self) -> str:
        return self._gob.getResname()

    ## Get name of window for gob from gobWindowMap
    # @return Window name
    def window_name_for_gob(self) -> str:
        return self._gob.windowNameForGob()

    ## Set gob as colored, replaces previous if color already set
    # @param r Red between 0-255
    # @param g Green between 0-255
    # @param b Blue between 0-255
    # @param a Alpha between 0-255
    def set_marked(self, r: int = 255, g: int = 0, b: int = 0, a: int = 255):
        self._gob.setMarked(r, g, b, a);

    ## Remove color marking
    def set_unmarked(self):
        self._gob.setUnmarked()

    ## Check if the object is moving
    # @return retuns true if gob is moving, false otherwise
    def is_moving(self) -> bool:
        return self._gob.isMoving()

    ## Get speed of this gob if it is moving
    # @return speed of gob
    def get_speed(self) -> float:
        return self._gob.getSpeed()

    ## Returns resnames of poses of this gob if it has any poses
    # @return resnames of poses
    def get_poses(self) -> List[str]:
        return list(self._gob.getPoses())

    ## Get overlays of the gob, get meshid with get_overlay_id
    # @return resnames of poses
    def get_overlay_names(self) -> List[str]:
        return list(self._gob.getOverlayNames())

    ## Return meshid of the overlay with the given resname
    # @param overlay_name exact match
    # @return meshid of the overlay or -1 if not found
    def get_overlay_id(self, overlay_name: str) -> int:
        return self._gob.getOverlayId(overlay_name)

    ## Sdt may tell information about things such as tanning tub state, crop stage etc.
    # @return sdt of the gob, -1 if not found
    def get_sdt(self, idx: int = 0) -> int:
        return self._gob.getSdt(idx)

    ## Get bounding boxes of the gob if some are found
    # @return List containing polygons which contain each point of the polygon
    def get_boundingbox(self) -> List[List[Tuple[float, float]]]:
        bb = self._gob.getBb()
        if bb is None:
            return []
        ret = []
        for pol in bb.polygons:
            ret.append([])
            for vert in pol.vertices:
                ret[len(ret)-1].append((vert.x, vert.y))
        return ret

    ## Get rectangle that covers bounding boxes of the gob if some are found
    # @return Corners of the rectangle
    def get_boundingbox_rect(self) -> Tuple[Tuple[float, float], Tuple[float, float]]:
        bb = self._gob.getBb()
        ret = [[-1e9,-1e9], [1e9,1e9]]
        if bb is None:
            return ((0, 0), (0, 0))
        for pol in bb.polygons:
            for vert in pol.vertices:
                ret[0][0] = max(ret[0][0], vert.x)
                ret[1][0] = min(ret[1][0], vert.x)
                ret[0][1] = max(ret[0][1], vert.y)
                ret[1][1] = min(ret[1][1], vert.y)
        return tuple([tuple(ret[0]), tuple(ret[1])])

    ## Get mapgrid id of the gobs current location
    def get_grid_id(self) -> int:
        return self._gob.getGridId()

    ## Get persistent location which can be translated back to coords with PBotUtils getcoords
    def get_persistent_loc(self) -> Tuple[int, float, float]:
        return (int(self.get_grid_id()), *tuple(float(c % (11*100)) for c in self.get_coords()))

    ## Check if the gob is KO/dead
    # @return true if the animal is knocked out, false if not
    def is_knocked(self) -> bool:
        return self._gob.isKnocked()

    ## Get list of set composite equs
    # @return list of resnames of equ, if has any
    def get_comp_nequ(self) -> List[str]:
        return list(self._gob.getCompEqu())

    ## Get list of set composite mods
    # @return list of resnames of mod, if has any
    def get_comp_nmod(self) -> List[str]:
        return list(self._gob.getCompMod())
